hd-wallet/plugin.lua

```diff
@@ -1,440 +1,418 @@
+    .. blob
+    .. Blob.from_hex("A00706052B8104000A")
+
+  return assert(Sobject.import {
+    obj_type       = "EC",
+    elliptic_curve = "SecP256K1",
+    value          = asn1,
+    transient      = true
+  })
 end
 
--- deserialize bip32 key
-function deserialize(exported_master_key_serialized)
-    local hex_key = Blob.from_base58(exported_master_key_serialized):hex()
-
-    key.version = string.sub(hex_key, 1, 8)
-    key.depth = string.sub(hex_key, 9, 10)
-    key.index = string.sub(hex_key, 11, 18)
-    key.fingerprint = string.sub(hex_key, 19, 26)
-    key.chain_code = string.sub(hex_key, 27, 90)
-    key.key = string.sub(hex_key, 91, 156)
-    key.checksum = string.sub(hex_key, 157, 164)
-
+-- RIPEMD-160(SHA-256(data))
+local function hash_160(data)
+  local sha256_hash = assert(digest {
+    data = data,
+    alg  = 'SHA256'
+  }).digest
+  local ripemd160_hash = assert(digest {
+    data = sha256_hash,
+    alg  = 'RIPEMD160'
+  }).digest
+
+  return ripemd160_hash
+end
 
-    if key.version ~= PRIVATE_WALLET_VERSION and key.version ~= PUBLIC_WALLET_VERSION then
-        error("Unexpected key version")
-    end
+---------- @@@@@@@@@@@@@@@@@@@@@@@@@ ----------
+----------           BIP 32          ----------
+---------- @@@@@@@@@@@@@@@@@@@@@@@@@ ----------
 
-    return key
+-- Named "point(p)" in the spec
+local function public_blob_from_private_blob(blob)
+  assert(#blob == 32)
+  local sobject = import_ec_key(blob)
+  local asn1_public_key = sobject.pub_key
+
+  -- extract coordinates from ASN1 public key
+  -- first half of last 64 bytes is x-coordinate and second half is y-coordinate
+  assert(#asn1_public_key, 88)
+  local x = asn1_public_key:slice(25, 25 + 32 - 1)
+
+  -- Point compression: If y is even, return 0x02 || x. Else, return 0x03 || x.
+  -- (See SEC1 sec. 2.3.3)
+  local one = Blob.from_hex("01")
+  local y_last_byte = asn1_public_key:slice(88, 88)
+  if y_last_byte & one == one then
+    return Blob.from_hex("03") .. x
+  end
+  return Blob.from_hex("02") .. x
 end
 
--- import ec key into sdkms
--- this will help to evaluate public key from private key
-function import_ec_key(blob)
-    local sobject = assert(Sobject.import { name = "ec", obj_type = "EC", elliptic_curve = "SecP256K1", value = blob, transient = true })
-    return sobject
+local function ser_32(i)
+  return Blob.from_hex(string.format("%08X", i))
 end
 
--- export secret key that holds the BIP32 master key --
--- export BIP32 master key from SDKMS ---
-function export_secret_key(keyId)
-    return Sobject { kid = keyId }:export().value
-end
+local function deserialize_bip32(blob)
+  local key = {
+    ["version"]     = blob:slice(1, 4),   --  4 bytes
+    ["depth"]       = blob:slice(5, 5),   --  1 byte
+    ["parent_fgpt"] = blob:slice(6, 9),   --  4 bytes
+    ["index"]       = blob:slice(10, 13), --  4 bytes
+    ["chain_code"]  = blob:slice(14, 45), -- 32 bytes
+    -- ignored                            --  1 0x00 byte
```

Comment:
This `0x00` byte was [hacked in](https://github.com/fortanix/sdkms-plugin-library/pull/8/files#diff-99dc3f06dd4742e3f7d914953fbdc77546009acc6f3b1896beea3495bd7812fcL249). It is simpler to just ignore it and comply with it on serialization.